### PR TITLE
Redesign `hex::BufEncoder` to accept owned arrays

### LINKED ISSUE #1270: `BufEncoder` trouble
I wanted to change #765 to use the newly added encoder but found out I made a design mistake. It's not possible to create an owned `BufEncoder`. I went ahead to redesign it but hit another design issue: `OutBytes` can not be created from `&T` making `BufEncoder::as_ref(&self)` impossible. Note that either way I have to add a special trait for converting to `OutBytes`. `From` is no longer enough because there's additional requirement for the returned reference to not depend on external state.

I see these options:

* Change the type to unsized and just use normal references, very clean but requires `unsafe` which we already determined is OK in other parts of the code.
* Make an immutable version of `OutBytes` with some conversions - annoying, adding another public type for the user to think about (even though it should be irrelevant)
* Make `as_str` and `is_full` take `&mut self` - this looks silly but probably not too limiting
* Make the special trait empty and have a method on the sealed one. Possibly use a bespoke trick to prevent calling the hidden method (not sure if it'll work). This avoids concerning users with another public item but contains hidden implementation detail as `pub` items.

Any more ideas or view which is the best? I personally like the first and last option.

### PR DESCRIPTION
**Pull request by Kixunil**
Not being able to create an owned `BufEncoder` prevented returning it from functions which need to allocate the buffer on stack. Such is the case in WIP serde via consensus serialzation.

This change refactors `OutBytes` to be unsized, adds an `AsOutBytes` trait and uses that one instead of `Into` to perform the conversion.

Closes #1270 

This is meant as potentially mergeable demonstration. Interestingly, it was easier than I expected.

### CODE REVIEW

**Comment by apoelstra** [comment]
This is really elegant! Out of curiosity where in std did you copy this from?

concept ACK, even with the unsafety.

Regarding xargo/miri, yes it's worth it!! We should MIRI-check a ton of stuff (annoyingly last I looked, anything that called into libsecp we couldn't do with miri which eliminates a lot of crypto..). But we should do this in a separate PR (even though initially they'll be "dumb" tests that don't even involve unsafety so theoretically miri should never get mad about).

**Comment by Kixunil** [comment]
Same as before: [`std::path::Path::new`](https://github.com/rust-lang/rust/blob/c97922dca563cb7f9385b18dbf7ca8c97f8e1597/library/std/src/path.rs#L1976)

`miri` is only useful for `unsafe` (as you noted `libsecp256k1` where it's unusable) and potentially for dependencies with `unsafe` that don't run it themselves or have poor coverage (can we avoid such dependencies?)

I'll try to add it.

File: internals/src/hex/buf_encoder.rs
Comment (Kixunil) [inline internals/src/hex/buf_encoder.rs:28]: Note one would normally have two traits - one for `mut` the other for non-`mut`. The limitation is that immutable version of `AsOutBytes` can not be implemented for `&T where T: AsOutBytes`.

I think this is fine, we really don't want to deal with immutable slices beyond providing non-weird API of `BufEncoder`

**Comment by Kixunil** [comment]
I propose to add `miri` tests after #1272 is merged which adds the test script. There shouldn't be much harm merging this sooner and it'd unblock my other work. But I can wait as well.

**Comment by apoelstra** [comment]
Sounds good.

re-concept ACK -- I had an idea that I could evade the unsafety by clever use of `AsRef<[u8]>` and `AsMut<[u8]>` but it didn't pan out. FYI the comment on the `OutBytes` type still says "we are trying to avoid unsafety for as long as possible". I'm not sure if this is stale or if it's meant to specifically refer to the use of `MaybeUninit`.

**Review by apoelstra — APPROVED** [review]
ACK 9a29b5b80a16e1042c74c40471cf2f2ae0eb86a1

**Comment by Kixunil** [comment]
We can't use `AsRef` or any other `core` trait because none of them guarantees pure functions. (Goes into the collection of traits that should be in `core`.) Without this change to `MaybeUninit` would be unsound and we'd have to break the API to fix it - what my design avoids.

Indeed the `unsafe` refers to `MaybeUninit` which is much more tricky than those two casts. (Requires `Out` type for instance.)

File: internals/src/hex/buf_encoder.rs
Comment (tcharding) [inline internals/src/hex/buf_encoder.rs:34]: I don't understand this comment (and module use), the `OutBytes` inner array is private so cannot be accessed by other modules anyways? Or am I missing something? I would have expected to see `OutBytes` as part of `buf_encoder` module and `private::Sealed`.

**Review by tcharding — APPROVED** [review]
Except for the question about the `out_bytes` module, ACK 1bf885550e773911a5a6ecbcf496264ad3a0db4b

File: internals/src/hex/buf_encoder.rs
Comment (Kixunil) [reply internals/src/hex/buf_encoder.rs:34]: If it was in `buf_encoder` it'd make the field accessible to the rest of the module which we don't want because it raises auditing cost. This is not critical now but if we ever switch to `MaybeUninit` we won't forget about it and will have a much smaller diff. :)

Comment (tcharding) [reply internals/src/hex/buf_encoder.rs:34]: Cool, can we change it to "This prevents the rest of the module from accessing the field of `OutBytes`" please.

Comment (tcharding) [reply internals/src/hex/buf_encoder.rs:34]: nit only, of course :)

Comment (Kixunil) [reply internals/src/hex/buf_encoder.rs:34]: and the crate :)

I'd prefer not to mess with a PR that has two ACKs for something this trivial.

**Review by apoelstra — APPROVED** [review]
ACK 1bf885550e773911a5a6ecbcf496264ad3a0db4b

**Pull request merged**
